File: Examenes-EDD-1/2020-Febrero/Java/src/dataStructures/graph/FordFulkerson.java
/**
 * Student's name: Paco Molina Cuenca
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 */

package dataStructures.graph;

import dataStructures.list.LinkedList;
import dataStructures.list.List;
import dataStructures.set.HashSet;
import dataStructures.set.Set;
import dataStructures.tuple.Tuple2;

public class FordFulkerson<V> {
    private WeightedDiGraph<V,Integer> g; // Initial graph 
    private List<WDiEdge<V,Integer>> sol; // List of edges representing maximal flow graph
    private V src; 			  // Source
    private V dst; 		  	  // Sink
	
    /**
     * Constructors and methods
     */

    public static <V> int maxFlowPath(List<WDiEdge<V,Integer>> path) {
        int pesoMinimo = path.get(0).getWeight();

        for(WDiEdge<V, Integer> arco : path){
            if(arco.getWeight() < pesoMinimo)
                pesoMinimo = arco.getWeight();
        }

        return pesoMinimo;
    }

    public static <V> List<WDiEdge<V,Integer>> updateEdge(V x, V y, Integer p, List<WDiEdge<V,Integer>> edges) {
        List<WDiEdge<V,Integer> resultado = new LinkedList<>();
        int nuevoPeso;

        for(WDiEdge<V,Integer> arco : edges){
            if(arco.getSrc().equals(x) && arco.getDst().equals(y)){
                nuevoPeso = arco.getWeight() + p;

                if(nuevoPeso != 0)
                    resultado.append(new WDiEdge<>(x, nuevoPeso, y));
            
            } else{
                resultado.append(arco);
            }
        }

        return resultado;

    }

    public static <V> List<WDiEdge<V,Integer>> updateEdges(List<WDiEdge<V,Integer>> path, Integer p, List<WDiEdge<V,Integer>> edges) {
        List<WDiEdge<V,Integer>> resultado = edges;

        for(WDiEdge<V, Integer> arco : path)
            resultado = updateEdge(arco.getSrc(), arco.getDst(), p, resultado);
        
        return resultado;
    }

    public static <V> List<WDiEdge<V,Integer>> addFlow(V x, V y, Integer p, List<WDiEdge<V,Integer>> sol) {
        List<WDiEdge<V,Integer> resultado = new LinkedList<>();

        for(WDiEdge<V,Integer> arco : sol){
            if(arco.getSrc().equals(x) && arco.getDst().equals(y))
                resultado.append(new WDiEdge<>(x, arco.getWeight() + p, y));
            else if(arco.getSrc().equals(y) && arco.getDst().equals(x) && arco.getWeight().equals(p))
                // Nada
            else if(arco.getSrc().equals(y) && arco.getDst().equals(x) && arco.getWeight() < p)
                resultado.append(new WDiEdge<>(x, p - arco.getWeight(), y));
            else if(arco.getSrc().equals(y) && arco.getDst().equals(x) && arco.getWeight() > p)
                resultado.append(new WDiEdge<>(y, arco.getWeight() - p, x));
            else 
                resultado.append(new WDiEdge<>(x, p, y));
        }

        return resultado;
    }

    public static <V> List<WDiEdge<V,Integer>> addFlows(List<WDiEdge<V,Integer>> path, Integer p, List<WDiEdge<V,Integer>> sol) {
        List<WDiEdge<V,Integer>> resultado = sol;

        for(WDiEdge<V, Integer> arco : path)
            resultado = addFlow(arco.getSrc(), arco.getDst(), p, resultado);
        
        return resultado;
    }

    public FordFulkerson(WeightedDiGraph<V,Integer> g, V src, V dst) {
        this.sol = new LinkedList<>();
        this.src = src;
        this.dst = dst;
        WeightedDiGraph<V,Integer> wdg = g;

        int mf;
        List<WDiEdge<V,Integer>> edges;
        WeightedBreadthFirstTraversal<WDiEdge<V,Integer>> pathBeginningInSrc = new WeightedBreadthFirstTraversal<>(g, src);
        List<WDiEdge<V,Integer>> pathFromSrcToDst = pathBeginningInSrc.pathTo(dst);

        while(pathFromSrcToDst != null && !pathFromSrcToDst.isEmpty()){
            mf = maxFlowPath(pathFromSrcToDst);
            edges = wdg.wDiEdges();
            edges = updateEdges(pathFromSrcToDst, -mf, edges);
            edges = updateEdges(reversePath(pathFromSrcToDst), mf, edges);
            wdg = new WeightedDictionaryDiGraph(wdg.vertices(), edges);
            this.sol = addFlows(pathFromSrcToDst, mf, sol);

            pathBeginningInSrc = new WeightedBreadthFirstTraversal<>(wdg, src);
            pathFromSrcToDst = pathBeginningInSrc.pathTo(dst);
        }

        this.g = wdg;
    }

    private List<WDiEdge<V,Integer>> reversePath(List<WDiEdge<V,Integer>> path){
        List<WDiEdge<V,Integer>> res = new LinkedList<>();

        for(WDiEdge<V,Integer> arco : path)
            res.append(new WDiEdge<>(arco.getDst(), arco.getWeight(), arco.getSrc()));

        return res;
    }

    public int maxFlow() {
        int flujoMax = 0;

        for(WDiEdge<V,Integer> arco : sol){
            if(arco.getSrc().equals(src))
                flujoMax += arco.getWeight();
        }

        return flujoMax;
    }

    public int maxFlowMinCut(Set<V> set) {
        int flujoMax = 0;

        for(WDiEdge<V,Integer> arco : g.wDiEdges()){
            if(set.isElem(arco.getSrc()) && !set.isElem(arco.getDst()))
                flujoMax += arco.getWeight();
            else if(!set.isElem(arco.getSrc()) && set.isElem(arco.getDst()))
                flujoMax -= arco.getWeight();
        }

        return flujoMax;
    }

    /**
     * Provided auxiliary methods
     */
    public List<WDiEdge<V, Integer>> getSolution() {
        return sol;
    }
	
    /**********************************************************************************
     * A partir de aquí SOLO para estudiantes a tiempo parcial sin evaluación continua.
     * ONLY for part time students.
     * ********************************************************************************/

    public static <V> boolean localEquilibrium(WeightedDiGraph<V,Integer> g, V src, V dst) {
        boolean hayEquilibrioTotal = true;

        for(V vertice : g.vertices()){
            if(!vertice.equals(src) && !vertice.equals(dst) && g.inDegree(vertice) != g.outDegree(vertice))
                hayEquilibrioTotal = false;
        }

        return hayEquilibrioTotal;
    }

    public static <V,W> Tuple2<List<V>,List<V>> sourcesAndSinks(WeightedDiGraph<V,W> g) {
        Tuple2<List<V>,List<V>> fuentesYSumideros = new Tuple2<>(new LinkedList<>(), new LinkedList<>());

        for(V vertice : g.vertices()){
            if(g.inDegree(vertice) == 0)
                fuentesYSumideros._1().append(vertice);
            else if(g.outDegree(vertice) == 0)
                fuentesYSumideros._2().append(vertice);
        }

        return fuentesYSumideros;
    }

    public static <V> void unifySourceAndSink(WeightedDiGraph<V,Integer> g, V newSrc, V newDst) {
        Tuple2<List<V>,List<V>> fuentesYSumideros = sourcesAndSinks(g);

        if(fuentesYSumideros._1().size() > 1){
            V nuevaFuente = newSrc;
            g.addVertex(nuevaFuente);

            for(V vertice : fuentesYSumideros._1())
                g.addDiEdge(nuevaFuente, pesoFuente(vertice, g), vertice);
        }

        if(fuentesYSumideros._2().size() > 1){
            V nuevoSumidero = newDst;
            g.addVertex(nuevoSumidero);

            for(V vertice : fuentesYSumideros._2())
                g.addDiEdge(vertice, pesoSumidero(vertice, g), nuevoSumidero);
        }
    }

    private static <V> int pesoFuente(V vertice, WeightedDiGraph<V,Integer> g){
        int sumaPesos = 0;

        for(WDiEdge<V,Integer> arco : g.wDiEdges()){
            if(arco.getSrc().equals(vertice))
                sumaPesos += arco.getWeight();
        }

        return sumaPesos;
    }

    private static <V> int pesoSumidero(V vertice, WeightedDiGraph<V,Integer> g){
        int sumaPesos = 0;

        for(WDiEdge<V,Integer> arco : g.wDiEdges()){
            if(arco.getDst().equals(vertice))
                sumaPesos += arco.getWeight();
        }

        return sumaPesos;
    }
}

File: Examenes-EDD-1/2020-Febrero/Haskell/DataStructures/Graph/FordFulkerson.hs
-------------------------------------------------------------------------------
-- Ford-Fulkerson Algorithm. Maximal flow for a weighted directed graph.
--
-- Student's name: Paco Molina Cuenca
-- Student's group: Doble Grado Matemáticas e Ingeniería Informática D
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.FordFulkerson where

import Data.List  ((\\))
import DataStructures.Graph.WeightedDiGraph
import DataStructures.Graph.WeightedDiGraphBFT

maxFlowPath :: Path (WDiEdge a Integer) -> Integer
maxFlowPath lista = minimum (map weightEdge lista)
    where 
        weightEdge :: WDiEdge a Integer -> Integer 
        weightEdge (E a peso b) = peso

updateEdge ::(Eq a) => a -> a -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
updateEdge x y p = map actualizaArco 
    where 
        actualizaArco e@(E o peso d)
            | o == x && d == y = E x p y
            | otherwise = e

updateEdges :: (Eq a) => Path (WDiEdge a Integer) -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
updateEdges [] _ edges = edges 
updateEdges (e@(E o p d):es) peso edges = updateEdges es peso (updateEdge o d peso edges)

addFlow :: (Eq a) => a -> a -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
addFlow x y p sol 
    | wInv /= 0 = updateEdge x y (p + wInv) sol 
    | w /= 0 && w == p = deleteEdge y x sol 
    | w /= 0 && w < p = (E x (p - w) y) : (deleteEdge y x sol)
    | w /= 0 && w > p = updateEdge y x (w - p) sol 
    | otherwise = (E x p y) : sol
    where 
        w = existeArco y x sol
        wInv = existeArco x y sol

        existeArco :: (Eq a) => a -> a -> [WDiEdge a Integer] -> Integer
        existeArco _ _ [] = 0 
        existeArco x y (e@(E o peso d):es)
            | x == o && y == d = peso
            | otherwise = existeArco x y es
        
        deleteEdge :: (Eq a) => a -> a -> [WDiEdge a Integer] -> [WDiEdge a Integer]
        deleteEdge x y [] = []
        deleteEdge x y (e@(E o _ d):es)
            | x == o && y == d = es 
            | otherwise = e : (deleteEdge x y es)

addFlows :: (Eq a) => Path (WDiEdge a Integer) -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
addFlows [] _ sol = sol 
addFlows (e@(E x peso y):es) p sol = addFlows es p (addFlow x y p sol) 

fordFulkerson :: (Ord a) => (WeightedDiGraph a Integer) -> a -> a -> [WDiEdge a Integer]
fordFulkerson g src dst = fordFulkersonRec [] g 
    where 
        fordFulkersonRec sol wdg = 
            case bftPathTo wdg src dst of 
                Nothing     -> sol 
                Just path   -> fordFulkersonRec sol' wdg'
                    where 
                        mf = maxFlowPath path
                        edges = weightedDiEdges wdg
                        edges' = updateEdges path (-mf) edges
                        edges'' = updateEdges (reversePath path) mf edges'
                        wdg' = mkWeightedDiGraphEdges (vertices wdg) edges''
                        sol' = addFlows path mf sol

                        reversePath :: Path (WDiEdge a Integer) -> Path (WDiEdge a Integer)
                        reversePath [] = []
                        reversePath (e@(E x peso y):es) = (E y peso x) : (reversePath es)

maxFlow :: (Ord a) => [WDiEdge a Integer] -> a -> Integer
maxFlow sol src = flujoMaximoRec sol 
    where 
        flujoMaximoRec [] = 0 
        flujoMaximoRec (e@(E x p y):es)
            | x == src = p + (flujoMaximoRec es)
            | otherwise = flujoMaximoRec es

maxFlowMinCut :: (Ord a) => (WeightedDiGraph a Integer) -> a -> a -> [a] -> Integer
maxFlowMinCut wdg src dst set = sum [p | E x p y <- aristas, x `elem` set, y `notElem` set] - sum [p | E x p y <- aristas, y `elem` set, x `notElem` set]
    where 
        aristas = weightedDiEdges wdg

-- A partir de aquí hasta el final
-- SOLO para alumnos a tiempo parcial 
-- sin evaluación continua

localEquilibrium :: (Ord a) => WeightedDiGraph a Integer -> a -> a -> Bool
localEquilibrium wdg src dst = comprobacionEquilibrioRec (vertices wdg)
    where 
        comprobacionEquilibrioRec [] = True 
        comprobacionEquilibrioRec (v:vs) 
            | v == src || v == dst = comprobacionEquilibrioRec vs
            | otherwise = (sum [p | E x p y <- aristas, x == v] == sum [p | E x p y <- aristas, y == v]) && comprobacionEquilibrioRec vs
                where 
                    aristas = weightedDiEdges wdg

sourcesAndSinks :: (Eq a) => WeightedDiGraph a b -> ([a],[a])
sourcesAndSinks wdg = ([v | v <- vertices wdg, inDegree wdg v == 0], [v | v <- vertices wdg, outDegree wdg v == 0])

unifySourceAndSink :: (Eq a) => WeightedDiGraph a Integer -> a -> a -> WeightedDiGraph a Integer
unifySourceAndSink wdg newSrc newDst
    | length fuentes == 1 && length sumideros == 1 = wdg 
    | length fuentes > 1 && length sumideros == 1 = mkWeightedDiGraphEdges (newSrc : (vertices wdg)) ((addFuentes fuentes) ++ (weightedDiEdges wdg))
    | length fuentes == 1 && length sumideros > 1 = mkWeightedDiGraphEdges (newDst : (vertices wdg)) ((addSumideros sumideros) ++ (weightedDiEdges wdg))
    | otherwise = mkWeightedDiGraphEdges (newSrc : newDst : (vertices wdg)) ((addFuentes fuentes) ++ (addSumideros sumideros) ++ (weightedDiEdges wdg))
    where 
        (fuentes, sumideros) = sourcesAndSinks wdg

        addFuentes [] = []
        addFuentes (f:fs) = (E newSrc peso f) : (addFuentes fs)
            where 
                peso = sum [p | E x p _ <- weightedDiEdges wdg, x == f]

        addSumideros [] = []
        addSumideros (s:ss) = (E s peso newDst) : (addSumideros ss)
            where 
                peso = sum [p | E _ p y <- weightedDiEdges wdg, y == s]
